File: 139_Word_Break.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        # Edge cases
        if(len(s) == 0):
            return True
        if(not wordDict):
            return False
        
        cache = [None] * len(s)
        def canSegment(i: int):
            # Base case
            if(i == len(s)):
                return True
            
            # Lookup result
            if(cache[i] is not None):
                return cache[i]
            
            # Recursively segment the characters, trying each word 
            # in wordDict at this pass
            for word in wordDict:
                nextIndex = i + len(word)
                if(nextIndex <= len(s) and s[i:nextIndex] == word and canSegment(nextIndex)):
                    cache[i] = True
                    return True
            
            # Was unable to segment the remaining indices using wordDict
            cache[i] = False
            return False
        
        return canSegment(0)
        

File: 139_Word_Break.java
// Assumption: s and wordDict only contain lowercase alpha characters

import java.util.*;

class Solution {
    public boolean beginsWith(String s, String word){
        if(s.length() < word.length())
            return false;
        
        int left = 0;
        int right = word.length() - 1;
        while(left <= right){
            if(s.charAt(left) != word.charAt(left) || s.charAt(right) != word.charAt(right))
                return false;
            left++;
            right--;
        }
        return true;
    }
    
    public boolean wordBreak(String s, List<String> wordDict) {
        // BFS traversal
        Queue<String> q = new LinkedList<String>();
        q.add(s);
        HashSet<String> visited = new HashSet<String>();

        String curr, remaining;
        while(!q.isEmpty()){
            curr = q.poll();
            if(visited.contains(curr))
                continue;
            visited.add(curr);
            
            // Expand each pathway of selected words
            for(String word : wordDict){
                if(beginsWith(curr, word)){
                    remaining = curr.substring(word.length());
                    if(remaining.length() == 0){                    // End of string reached
                        return true;
                    }
                    q.add(remaining);
                }
                
            }
        }
        return false;
    }
}